linsolve: add GMRES

vladimir-ch opened the pull request.

vladimir-ch commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,288 @@
+	av []float64
+
+	// v is an n×(m+1) matrix (where m == Restart)
+	// in column-major storage. Its columns form
```
This has been my main dilemma. Mathematics stores vectors in the columns of a matrix. All GMRES descriptions follow this. I'd like to avoid strided copies between columns of V and Context.Src and Dst. Alternatively, we can say that we are storing the vectors in rows and that our V and H are V^T and H^T in a standard description of GMRES.

btracey replied:
I think that's probably better than the column major storage. I would call the fields `Vt` and `Ht` and comment that we are storing the transpose so that vectors are rows.

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,288 @@
+	"gonum.org/v1/gonum/floats"
+)
+
+// GMRES implements the Generalized Minimum Residual method with the modified
```
s/with the modified/with modified

vladimir-ch replied:
I kept `with the modified Gram-Schmidt`, I think it is actually an established name for the algorithm. Let me know if my justification is wrong (outside of textbook examples the rules for using "the", "a" or nothing still seem rather esoteric to me).

btracey replied:
If it's an established name, that's fine. 

Okay, so I wrote half of a paragraph, and realized I can't come up with the rule on "the" vs. no "the". The fundamental tension, I believe, is the multiple meanings of the word "with", and whether it means "using" or whether it means "that has", in other words, whether what comes after the "with" modifies the noun or the verb. So, we could say that "package linsolve solves systems of linear equations with an iterative method". Here "with" is a synonym for "using", it is how we are solving the system. On the other hand, we may have "Halton implements the Halton sequence with scrambling". In this case, it's not saying how we are implementing the Halton sequence, it's saying we're implementing a specific kind of Halton sequence (that uses scrambling). So, in the case here, it reads to me as "using the Gram-Schmidt", whereas I think it means "the kind of GMRES that has modified Gram-Schmidt".

On the other hand, I can think of obvious counterexamples to my argument above, so it's certainly not a strong rule. English is dumb.

kortschak replied:
Articles are the greatest shibboleth.

btracey replied:
And prepositions

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,288 @@
+)
+
+// GMRES implements the Generalized Minimum Residual method with the modified
+// Gram-Schmidt orthogonalization  for solving systems of linear equations
```
remove extra space

vladimir-ch replied:
Done. Thanks.

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,288 @@
+// References:
+//  - Barrett, R. et al. (1994). Section 2.3.4 Generalized Minimal Residual (GMRES).
+//    In Templates for the Solution of Linear Systems: Building Blocks for
+//    Iterative Methods (2nd ed.) (pp. 12-15). Philadelphia, PA: SIAM.
```
This says p12-15, but the actual GMRES algorithm is on page 18

vladimir-ch replied:
Thanks, fixed.

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,288 @@
+	// Restart is the restart parameter which
+	// limits the computation and storage costs.
+	// It must hold that
+	//  0 <= Restart <= n
```
Is it the case that really 1 <= Restart <= n? I think that's more clear for the comment if so (a valid value is in this range, but the default value gives you n).

vladimir-ch replied:
Yes, `1 <= Restart <= n` is correct but I didn't want to exlude 0 if it's actually allowed as a placeholder for a default value.

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,288 @@
+	// It must hold that
+	//  0 <= Restart <= n
+	// where n is the dimension of the problem.
+	// If it is 0, it will be set to n. Note
```
I don't like the way this comment reads. It sounds like it says "The default value of the parameter is a bad idea"

I'm not an expert in GMRES, but it sounds like a better default is hard to choose because of convergence guarantees. I would instead say something like `If Restart is 0, it will be set to n. This guarantees convergence of GMRES and increases robustness. Many specific problems however, particularly for large n, will benefit in efficiency by limiting the memory to a problem-dependent value less than n`

vladimir-ch replied:
I changed to what you suggest (although the part about a bad idea is in fact true for large n).

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,288 @@
+//    doi:10.6028/jres.049.044
+//    Retrieved from https://web.stanford.edu/class/cme324/saad-schultz.pdf
+type GMRES struct {
+	// Restart is the restart parameter which
```
Widen these comments out to 80 character lines.

vladimir-ch replied:
Done although I'm aware that @kortschak prefers field comments narrow.

kortschak replied:
The issue to consider here is how it is rendered in godoc, since we can't rely on godoc to do text flow for us with struct comments. In code I prefer narrow comments to ease reading effort (comments there are often more detailed or subtle and so this helps), these comments look OK at the longer length I think, but consider how it renders on a smaller screen via godoc.org.

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,288 @@
+	floats.Scale(1/wnorm, w) // Normalize v_{k+1}.
+}
+
+// qr uses Givens rotations to reduce H to upper-triangular form.
```
Is there something special about GMRES where we can't use the Lapack equivalent of this? For instance, Dgehrd?

vladimir-ch replied:
`Dgehrd` probably not, it works on a matrix as a whole while GMRES builds the factorization one column at each iteration. `Dlarf*` functions could be useful if we used the Householder reflectors to build the factorization. This is in fact an existing variant of GMRES that may be more stable but is also more expensive. I would like to implement it but it would probably deserve its own type.

btracey commented:
Higher level comment is that it would be best to use `mat` rather than `blas` directly. If we need to add to `mat` to make this possible, let's see if we can.

vladimir-ch commented:
PTAL

bq. Higher level comment is that it would be best to use mat rather than blas directly.

Yeah, probably it's because I feel more comfortable with BLAS. If I were to use it, I think the only missing piece would be getting a view of `Dense` as a triangular matrix. I'm not sure if it makes sense adding something like it to `mat`, it looks quite special. Passing it then to `SolveVec` should already work. The rest could be done with `RawRowView`.

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,287 @@
+		// Construct (k+1)-th column of the upper Hessenberg matrix H using
+		// the modified Gram-Schmidt process on V and v_{k+1} so that it is
+		// orthonormal to the previous k columns.
+		g.modifiedGS()
```
`g.modifiedGS(g.k, g.vt, g.hcol(k), ctx.Dst)`   See comment below.

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,287 @@
+		return PreconSolve, nil
+	case 5:
+		// v_{k+1} = M^{-1} * (A * v_k)
+		copy(g.vcol(g.k+1), ctx.Dst)
```
I think this copy is harmful for following the pseudocode in the templates. I'd rather the copy happen in `modifiedGS`

vladimir-ch replied:
If I do it in `modifiedGS`, then it does too many things and documenting it clearly is next to impossible. I really tried but I think it is better this way.

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,287 @@
+		// v_{k+1} = M^{-1} * (A * v_k)
+		copy(g.vcol(g.k+1), ctx.Dst)
+		// Construct (k+1)-th column of the upper Hessenberg matrix H using
+		// the modified Gram-Schmidt process on V and v_{k+1} so that it is
```
This is supposed to mean `and also compute v_{k+1} ...`, yes?

vladimir-ch replied:
Reworked the comment after changing `modifiedGS`.

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,287 @@
+}
+
+// modifiedGS applies the modified Gram-Schmidt procedure on the (k+1)-th row of V^T.
+func (g *GMRES) modifiedGS() {
```
It's good to have `modifiedGS` be a method for namespace reasons, but this is effectively a function to operate on some data, rather than really needing an object-oriented call. As such, I'd much rather this take in as arguments what it needs, and not reference the data in `GMRES` at all.

```
func (*GMRES) modifiedGS(k int, vt []float64, hk []float64, w []float64)
```

vladimir-ch replied:
Done.

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,287 @@
+}
+
+// qr uses Givens rotations to reduce H to upper-triangular form.
+func (g *GMRES) qr() {
```
Same comment as above for using input arguments.

vladimir-ch replied:
Done.

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,287 @@
+	}
+	wnorm := floats.Norm(w, 2)
+	hk[k+1] = wnorm          // H[k+1,k] = |v_{k+1}|
+	floats.Scale(1/wnorm, w) // Normalize v_{k+1}.
```
Then put the copy here  (in replace of before the call).

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,287 @@
+	g.y = reuse(g.y, dim)
+	g.av = reuse(g.av, dim)
+
+	ldv := dim
```
Do these also need to be zeroed? Looks like it.

vladimir-ch replied:
I added the zeroing to `reuse`.

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,287 @@
+	}
+}
+
+// drotg returns Givens plane rotation.
```
Use the BLAS function?

vladimir-ch replied:
Done.

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,287 @@
+	return givens{c: c, s: tmp * c}
+}
+
+// drot applies Givens rotation g to the vector [x,y] and returns the result.
```
Use the BLAS function?

vladimir-ch replied:
The BLAS operates on two vectors, not on a 2d vector. I renamed it to `applyGivens`

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,287 @@
+		// orthonormal to the previous k columns.
+		g.modifiedGS()
+		// Reduce H back to upper triangular form.
+		g.qr()
```
Is there a better description of the `J_1, ...`. It looks like what you're doing here, just curious if there's an easier way to check than me thinking through it.

btracey commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,287 @@
+		// Reduce H back to upper triangular form.
+		g.qr()
+
+		ctx.ResidualNorm = g.residualNorm()
```
Inputs here too I think.

vladimir-ch replied:
Changed the code here a bit.

btracey commented on linsolve/gmres_test.go:
```diff
@@ -0,0 +1,118 @@
+		randomSPD(50, rnd),
+		randomSPD(100, rnd),
+		randomSPD(200, rnd),
+		randomSPD(500, rnd),
```
Other random test matrices too since this works for more than SPD?

vladimir-ch replied:
Can I add these test later? I'd like to merge this first to see some motivating progress.

btracey replied:
Sure.

vladimir-ch commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,287 @@
+			g.givs[i].s = 0
+		}
+	}
+
```
I thought about it but it is already correctly reset before the inner iteration starts, so resetting it also here would be strange.

vladimir-ch commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,287 @@
+		panic("gmres: dimension not positive")
+	}
+
+	if g.Restart == 0 {
```
Done.

vladimir-ch commented on linsolve/gmres.go:
```diff
@@ -0,0 +1,287 @@
+// GMRES implements the Generalized Minimum Residual method with the modified
+// Gram-Schmidt orthogonalization for solving systems of linear equations
+//  A*x = b,
+// where A is a square, non-symmetric matrix. It uses restarts to control
```
Done.

vladimir-ch commented:
I really like how the implementation turned out. I think I addressed all your comments. PTAL @btracey

btracey commented:
I haven't had a chance to check the changes yet, FYI.

vladimir-ch commented:
Oh, sorry, I misunderstood the situation. Do you still want to go through it and add comments here?

btracey commented:
I think I do, I'm just really busy this week. I hadn't looked at the tests before if you want them checked.

The thing I note at the moment is that "ApplyGivens" is actually `blas.Drotm`, right?

vladimir-ch commented:
Both `Drot` and `Drotm` operate on two n-vectors. I think it makes little sense to create two slices with one element just to be able to call one of them.

vladimir-ch commented:
Actually, I ended up trying to use`Drot` and it looks much better then I expected. So I switched to `Drot` in 0c6f383c78ba17376e0c3c7f397e5c008bbc7f63 and removed `applyGivens`. Thanks for bringing it up.

You can take a look at the tests but just the code is enough. I want to return to the tests later, and extend them so that they use not only Matrix Market data.

btracey commented:
Just want to make sure I'm not holding anything up.

vladimir-ch commented:
Not at all. I'm slowly preparing a next PR with the driver function but I'm afraid that submitting it will have to wait until January.

btracey commented:
That's okay. I just realized we left it somewhat ambiguous so wanted to make sure you weren't waiting on me.
